### DISCUSSION TITLE
How to determine the number of DOF of a mechanism

### DISCUSSION THREAD

**Discussion by ManuelZ** [jgerstmayr/EXUDYN#42] [Q&A on usage of Exudyn] [answered]
I'm having trouble understanding if my 2D mechanism has 1 or 2 Degrees of Freedom. 

![image](https://user-images.githubusercontent.com/115771/232862756-62ea233a-6be2-4e45-a9c0-a9b674ff7990.png)

The software I'm using says that it has 2, but with a redundancy. It forces me to use two actuators to command motion. Grübler's criterion says it has 1 DOF, but I know that I can only trust Grübler into giving me a lower bound for the number of DOF.

Is there a way of calculating this with Exudyn?

As an additional question. I'm trying to learn some Screw theory to be able to apply some of the techniques that I'm seeing around (e.g. "A Unified Methodology for Mobility Analysis Based on Screw Theory") to be able to determine the mobility of my mechanism, and I noticed that Exudyn has nothing to do with this Screw theory. What would be the equivalent "methodology" used by Exudyn?

**Comment by jgerstmayr** [comment, marked as answer]
Thanks for your questions!

The problems you have are the following:
1) Grübler-Kutzbach always works if you exclude redundant constraints; your mechanism has 2 DOF, except for the singularities; you can move the two outmost supports independently
2) If you do not see how to compute Grübler-Kutzbach, you should not go into screw theory; it will not help you nor solve your problems
3) "Exudyn has nothing to do with this Screw theory": this is simply wrong. Screw theory is embedded, as all forces and torques and jacobians are computed accordingly, but it is not always stated that it follows from screw theory - like in most multibody software. The Lie group integrators and the geometrically exact beam theory very much uses these concepts as well. For a direct computation with the 6D vectors and matrices, there are some Python user functions denoted as T66 (Plücker) transformations, see the rigidBodyUtilities.py. But it is just a matter of notations, the results are always the same.

A) you cannot compute Grübler-Kutzbach directly, but you could do so by computing the jacobian and applying a singular value decomposition, which would give you a Nullspace of size 2. I will consider to make this as a function in the future.

**Comment by ManuelZ** [comment]
Thank you very much for your answers! Sorry for my wrong statement.

1) I will look into how to identify redundant constraints. I have been reading the Douglass Blanding's book called "Exact Constraint - Machine Design Using Kinematic Processing (1999)" and I find it very enlightening, but I still have practical doubts. Would there be any other reference out there from where to learn how to do this?

2) I have been trying to obtain the Jacobian through a function call but I haven't been able to find a way to do that. I have, though, been able to find a very big Jacobian-related matrix in the `solverInformation.txt` file, after activating the necessary options in `solutionSettings` through `SimulationSettings()`:

```
+++++++++++++++++++++++++++++++
EXUDYN V1.6.0 solver: implicit second order time integration
number of steps = 100000
...
    initial accelerations update Jacobian: Jac    = [[2499.99, ...],[0,2499.99,...],...]
initial accelerations = [...]
  Generalized-alpha: spectralRadius=0.9, alphaM=0.421053, alphaF=0.473684, beta=0.277008, gamma=0.552632, factJacA=0.909091
  has constant mass matrix=0
```

Is this what I need?

3) To manually obtain the Jacobian myself for the 2D and 3D case of this mechanism, I've been looking up into Chapter 7 of "Modern Robotics - Mechanics, Planning, and Control (2019)" (freely available [here](https://hades.mech.northwestern.edu/index.php/Modern_Robotics)), which has the most clear explanations that I have been able to find for closed-chain mechanisms, with specific examples of parallel mechanisms. The book uses Screw Theory though. Do you know any other reference out there?

**Comment by jgerstmayr** [comment]
To obtain the jacobian of your system, you have to use the solver (As static or implicit solvers anyway compute the jacobian).  See e.g. the function ComputeLinearizedSystem(..):
https://exudyn.readthedocs.io/en/latest/docs/RST/pythonUtilities/solver.html#function-computelinearizedsystem
and the source code when clicking on the function name. It shows how to compute the jacobian generally. I will add some functionality related to DOF counting soon.

**Comment by ManuelZ** [comment]
Thank you.

**Reply by jgerstmayr** [reply]
In the next update, there will be a function ComputeSystemDegreeOfFreedom in exudyn.solver. 
I assume that it will take a few days to see the pre-release for this update on github / pypi.
